test/upgrade/installation/shell.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package installation
+
+import (
+	"knative.dev/hack/shell"
+	pkgupgrade "knative.dev/pkg/test/upgrade"
+)
+
+// Base installs Operator with Knative Serving and Eventing from the previous stable release.
+func Base() pkgupgrade.Operation {
+	return install("OperatorWithServingEventingPreviousRelease", "install_previous_operator_release")
+}
+
+// LatestRelease installs Knative Serving and Eventing from the latest stable release.
+func LatestRelease() pkgupgrade.Operation {
+	return install("ServingEventingLatestRelease", "create_latest_custom_resource")
+}
+
+// PreviousRelease installs Knative Serving and Eventing from the previous stable release.
+func PreviousRelease() pkgupgrade.Operation {
+	return install("ServingEventingPreviousRelease", "install_previous_knative")
+}
+
+func install(installName, shellFunc string) pkgupgrade.Operation {
```

Comment:
I'm slightly terrified at how deep this goes between `go` and `bash`, but I don't have a well-considered alternative right now.

Reply:
Golang code calls the function in bash scripts.
There is sh file to import.
Based on knative's need, it calls the function to upgrade and downgrade, and the tests, after running them.

Reply:
The scary bit is going Bash -> Golang -> Bash in layers, makes it hard to reason about the test, iiuc.